File: app/src/main/java/net/twisterrob/colorfilters/android/matrix/RotatesComponent.java
package net.twisterrob.colorfilters.android.matrix;

import android.content.SharedPreferences;
import android.graphics.ColorMatrix;
import android.view.View;

class RotatesComponent extends Component {
    private final RotateComponent[] rotates;

    public RotatesComponent(View view, RefreshListener listener) {
        super(view, listener);
        rotates = new RotateComponent[]{
                new RotateComponent(view, listener, COMP_R),
                new RotateComponent(view, listener, COMP_G),
                new RotateComponent(view, listener, COMP_B)
        };
    }

    @Override
    public void setupUI() {
        for (Component rotate : rotates) {
            rotate.setupUI();
        }
    }


    @Override
    public void saveToPreferences(SharedPreferences.Editor editor) {
        for (Component rotate : rotates) {
            rotate.saveToPreferences(editor);
        }
    }

    @Override
    public void restoreFromPreferences(SharedPreferences prefs) {
        for (Component rotate : rotates) {
            rotate.restoreFromPreferences(prefs);
        }
    }

    @Override
    public void reset() {
        for (Component rotate : rotates) {
            rotate.reset();
        }
    }

    @Override
    public void refreshModel() {
        for (Component rotate : rotates) {
            rotate.refreshModel();
        }
    }

    @Override
    public void combineInto(ColorMatrix colorMatrix) {
        for (Component rotate : rotates) {
            rotate.combineInto(colorMatrix);
        }
    }

    @Override
    public boolean appendTo(StringBuilder sb) {
        boolean written = false;
        for (Component rotate : rotates) {
            written |= rotate.appendTo(sb);
        }
        return written;
    }

    public Component get(int compRGB) {
        return rotates[compRGB];
    }
}

File: feature/matrix/src/main/kotlin/net/twisterrob/colorfilters/android/matrix/RotatesComponent.kt
package net.twisterrob.colorfilters.android.matrix

import android.content.SharedPreferences
import android.graphics.ColorMatrix
import android.view.View

internal class RotatesComponent(
	view: View,
	listener: Component.RefreshListener
) : Component(view, listener) {

	private val rotates = arrayOf(
		RotateComponent(view, listener, COMP_R),
		RotateComponent(view, listener, COMP_G),
		RotateComponent(view, listener, COMP_B)
	)

	override fun setupUI() {
		rotates.forEach { it.setupUI() }
	}

	override fun saveToPreferences(editor: SharedPreferences.Editor) {
		rotates.forEach { it.saveToPreferences(editor) }
	}

	override fun restoreFromPreferences(prefs: SharedPreferences) {
		rotates.forEach { it.restoreFromPreferences(prefs) }
	}

	override fun reset() {
		rotates.forEach { it.reset() }
	}

	override fun refreshModel() {
		rotates.forEach { it.refreshModel() }
	}

	override fun combineInto(colorMatrix: ColorMatrix) {
		rotates.forEach { it.combineInto(colorMatrix) }
	}

	override fun appendTo(sb: StringBuilder): Boolean =
		rotates.fold(false) { written, component ->
			written or component.appendTo(sb)
		}

	operator fun get(compRGB: Int): Component =
		rotates[compRGB]
}
